### PR TITLE
[2829] Handle broken avatar image URLs

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://github.com/GetStream/stream-chat-android/issues/2829

### 🎯 Goal

Currently, we display nothing for a broken avatar image URL. We need to fall back to user initials in that case.

### 🛠 Implementation details

Avatars now show fallback initials in case there was an error while loading images from the network.

### 🎨 UI Changes

| Before | After |
| --- | --- |
| photo_2021-12-27 14 43 32 | photo_2021-12-27 14 43 28 |

**UserAvatar Previews:**
Screenshot 2021-12-27 at 14 18 05

**ChannelAvatar Previews:**

Screenshot 2021-12-27 at 14 17 54

**Avatar Previews:**

Screenshot 2021-12-27 at 14 18 24

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-team or #compose-chat-sdk-team) (required)
- [x] PR targets the `develop` branch

#### Code & documentation
- [x] Changelog is updated with client-facing changes
~~- [ ] New code is covered by unit tests~~
- [x] Comparison screenshots added for visual changes
~~- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)~~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt [hunk 1]
```diff
@@ -1,50 +1,128 @@
...
         )
     } else {
-        modifier
+        ImageAvatar(
+            modifier = modifier,
+            shape = shape,
+            painter = painter,
+            contentDescription = contentDescription,
+            onClick = onClick
+        )
```
Comment (skydoves) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt:78]: Just small a suggestion from an idea. 😄  

How about handling each case of the `ImagePainter.State` like the below?
Maybe we can consider receiving a `placeHolderPainter` parameter which is the type of the `Painter`, so clients can show a placeholder image when loading the images. (I guess if we're in a poor network environment, it could be taken a few seconds for loading an image.)

```kotlin
when (painter.state) {
        is ImagePainter.State.Error -> {
            InitialsAvatar(
                modifier = modifier,
                initials = initials,
                shape = shape,
                textStyle = textStyle,
                onClick = onClick
            )
        }
        is ImagePainter.State.Success, ImagePainter.State.Empty -> {
            ImageAvatar(
                modifier = modifier,
                shape = shape,
                painter = painter,
                contentDescription = contentDescription,
                onClick = onClick
            )
        }
        is ImagePainter.State.Loading -> {
            ImageAvatar(
                modifier = modifier,
                shape = shape,
                painter = placeHolderPainter,
                contentDescription = contentDescription,
                onClick = onClick
            )
        }
    }
```
Also, we may consider making it smooth by wrapping the above code with `Crossfade(targetState = painter.state)`.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt [hunk 2]
```diff
@@ -1,50 +1,128 @@
...
+        ImageAvatar(
+            modifier = modifier,
+            shape = shape,
+            painter = painterResource(id = R.drawable.stream_compose_preview_avatar),
+            contentDescription = contentDescription,
+            onClick = onClick
+        )
+        return
+    }
+    if (imageUrl.isNullOrBlank()) {
```
[the remark this answers is not in the document]
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt:50]: Also I think we can remove this - we should have a check in `UserAvatar` that does this.

To me, this logic/code should probably be there, and not in `Avatar`.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt (hunk 1 above)
Comment (filbabic) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt:78]: This is a great suggestion - if we're using this painter.state API, then we can do this and we can add the placeholderImage param that's empty or defined by us, by default.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/ImageAvatar.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.layout.ContentScale
+import io.getstream.chat.android.compose.ui.theme.ChatTheme
+
+/**
+ * Avatar/image that renders whatever image the [painter] provides. It allows for customization,
```
Comment (filbabic) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/ImageAvatar.kt:17]: I would rephrase the first sentence to:

`Renders an image the [painter] provides.`

**Review by filbabic — COMMENTED** [review]
Left a couple of comments, but overall this looks great! :]

I'll let you resolve those, but for now, the checklist is as follows:

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] Bugs validated (bugfixes)
- [x] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt (hunk 1 above)
Comment (bychkovdmitry) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/avatar/Avatar.kt:91]: Done, thanks for the suggestion!

**Review by filbabic — APPROVED** [review]
